### PR TITLE
[Merged by Bors] - tortoise: encode abstain votes as an array with layers

### LINKED ISSUE #2369: blocks, tortoise: merge for/against exceptions into a single list, vote neutral on layers instead of blocks
As described in https://github.com/spacemeshos/SMIPS/issues/2, the initial version of verifying tortoise includes three explicit lists of exception votes:

```
Block {
   Base_Block: <block_id>
   For_Exception_List: [<block_id>, <block_id>, <block_id>...] // List of blocks I vote For and Base_Block doesn't
   Against_Exception_List: [<block_id>, <block_id>, <block_id>...] // List of blocks I vote Against and Base_Block doesn't
   Neutral_Exception_List: [<block_id>, <block_id>, <block_id>...] // List of blocks I vote neutral and Base_Block doesn't or blocks that just come after Base_Block.
}
```

As discussed on the R&D call with Tal this week, we can optimize this in a few ways:

- blocks only ever vote neutral on an entire layer, never on individual blocks in a layer (and only in the case where they're still waiting for a Hare result for that layer). So there's no need to list individual block ids in the neutral list, just layer IDs.
- the For and Against exception lists can be collapsed down to a single list. For any block on this list, the block's vote is implicitly the _opposite_ of the base block's vote.

Regarding neutral votes in the base block, by default a block citing it should also be considered to vote neutral for the same layers, unless the block includes _any_ explicit for or against votes for _any_ blocks in that layer (this may already be how things work, I'm not sure)

CC @y0sher, @sudachen 

Addendum: a block does not need to encode an explicit vote _for_ its base block, this can be made automatic and implicit. ([discussion](https://spacemesh.slack.com/archives/C89QJJY3G/p1635846350093700))

### PR DESCRIPTION
**Pull request by dshulyak**
## Motivation

In verifying tortoise we should not count weight from the ballot if the ballot abstained for that layer. This will be easier to implement if abstain votes were encoded as an array of layer. 

Also, this change is a must for UBC, as we don't know the actual block value before hare termination. So there is no way to vote abstain.

Support/Against are not collapsed into a single list as it leads to less efficient implementation. Maybe we should revise it in the future.

Closes: https://github.com/spacemeshos/go-spacemesh/issues/2369

## Changes
- encode abstain votes as an array of layers

## Test Plan
existing unit tests, logically nothing has changed

### CODE REVIEW

**Review by countvonzero — APPROVED** [review]

File: tortoise/tortoise.go [hunk 1]
```diff
@@ -356,6 +356,10 @@ func (t *turtle) encodeVotes(
 				votes.Support = append(votes.Support, bid)
 			case against:
 				votes.Against = append(votes.Against, bid)
+			case abstain:
```
Comment (countvonzero) [inline tortoise/tortoise.go:360]: maybe instead comment that tortoise cannot vote abstain past hdist.

File: common/types/votes.go [hunk 2]
```diff
@@ -0,0 +1 @@
+package types
```
Comment (nkryuchkov) [inline common/types/votes.go]: Why is this file needed?

File: tortoise/tortoise.go [hunk 3]
```diff
@@ -744,6 +743,19 @@ func (t *turtle) addLocalVotes(ctx context.Context, logger log.Log, lid types.La
...
+	genesis := types.GetEffectiveGenesis()
+	limit := genesis
+	if state.last.After(genesis.Add(config.Zdist)) {
+		limit = state.last.Sub(config.Zdist)
+	}
+	_, isUndecided := state.undecided[lid]
+	if !lid.Before(limit) && isUndecided {
+		return true
+	}
+	return false
```
Comment (nkryuchkov) [inline tortoise/tortoise.go]:
```suggestion
	return !lid.Before(limit) && isUndecided
```

**Review by nkryuchkov — APPROVED** [review]

File: common/types/votes.go (hunk 2 above)
Comment (dshulyak) [reply common/types/votes.go]: my bad, forgot to remove it

File: tortoise/tortoise.go (hunk 1 above)
Comment (dshulyak) [reply tortoise/tortoise.go:360]: will change to an error, just in case

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge
